pkg/util/active_groups.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"go.uber.org/atomic"
+)
+
+const maxGroupsPerUser int = 1000
+
+type ActiveGroups struct {
+	mu                sync.RWMutex
+	timestampsPerUser map[string]map[string]*atomic.Int64 // map[user][group] -> timestamp
+}
+
+func NewActiveGroups() *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+	}
+}
+
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, ts int64) {
+	// Create new atomic before lock is acquired
+	newAtomic := atomic.NewInt64(ts)
```

Comment:
This allocates a new atomic even though the 99.99% percent of the requests will not need it. Should be allocated later when it's seen that it's _needed_, or at least _most likely needed_